### PR TITLE
local rate limiter: backport the non-timer based bucket

### PR DESCRIPTION
**Pull request by nfuden**
backport the first 3 rate limit pieces of upstream's 34774 issue

This allows us to run local rate limiting not on the main thread if an environment variable is set to true.

In the future can be disabled via envoy_reloadable_features_no_timer_based_rate_limit_token_bucket
for now can be ENABLED via envoy_reloadable_features_no_timer_based_rate_limit_token_bucket

### CODE REVIEW

**Comment by ashishb-solo** [comment]
I'm not sure exactly how to review/test this. So far, `envoy-static` seems to compile successfully. I'm also looking at the tests that were modified in the patch. It looks like here's a list of test files that were updated:

```
test/common/common/BUILD
test/common/common/token_bucket_impl_test.cc
test/extensions/filters/common/local_ratelimit/BUILD
test/extensions/filters/common/local_ratelimit/local_ratelimit_test.cc
test/extensions/filters/http/local_ratelimit/BUILD
test/extensions/filters/http/local_ratelimit/config_test.cc
test/extensions/filters/http/local_ratelimit/filter_test.cc
test/extensions/filters/http/local_ratelimit/local_ratelimit_integration_test.cc
test/extensions/filters/listener/local_ratelimit/local_ratelimit_test.cc
test/extensions/filters/network/local_ratelimit/local_ratelimit_integration_test.cc
test/extensions/filters/network/local_ratelimit/local_ratelimit_test.cc
```

So we can test this by running

bazel test \
@envoy//test/common/common:token_bucket_impl_test \
@envoy//test/extensions/filters/common/local_ratelimit/... \
@envoy//test/extensions/filters/http/local_ratelimit/... \
@envoy//test/extensions/filters/listener/local_ratelimit/... \
@envoy//test/extensions/filters/network/local_ratelimit/...

I'm in the process of running this locally. If it's possible to add this to CI, that would be amazing, but if it's too much of a lift, then don't worry about it.

I'm also wondering if we want to do some sort of performance test before merging this into Gloo and/or releasing it. Do you have any thoughts here?

**Comment by nfuden** [comment]
https://github.com/solo-io/gloo/pull/9822 shows that basic local rate limit works with this change

File: bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch [hunk 1]
```diff
@@ -0,0 +1,3422 @@
...
+ } // namespace Envoy
+diff --git a/source/common/runtime/runtime_features.cc b/source/common/runtime/runtime_features.cc
+index 260a08820c..f22d9cf58b 100644
+--- source/common/runtime/runtime_features.cc
++++ source/common/runtime/runtime_features.cc
+@@ -73,6 +73,7 @@ RUNTIME_GUARD(envoy_reloadable_features_lowercase_scheme);
+ RUNTIME_GUARD(envoy_reloadable_features_no_downgrade_to_canonical_name);
+ RUNTIME_GUARD(envoy_reloadable_features_no_extension_lookup_by_name);
+ RUNTIME_GUARD(envoy_reloadable_features_no_full_scan_certs_on_sni_mismatch);
++// RUNTIME_GUARD(envoy_reloadable_features_no_timer_based_rate_limit_token_bucket);
```
Comment (ashishb-solo) [inline bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch:375]: TODO remove this

File: bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch [hunk 2]
```diff
@@ -0,0 +1,3422 @@
...
+ RUNTIME_GUARD(envoy_reloadable_features_normalize_host_for_preresolve_dfp_dns);
+ RUNTIME_GUARD(envoy_reloadable_features_oauth_make_token_cookie_httponly);
+ RUNTIME_GUARD(envoy_reloadable_features_oauth_use_standard_max_age_value);
+@@ -113,6 +114,9 @@ RUNTIME_GUARD(envoy_restart_features_use_fast_protobuf_hash);
+ 
+ // Begin false flags. Most of them should come with a TODO to flip true.
+ 
++//backporting this so TODO: Remove all this forked code before adding to 1.31
```
Comment (ashishb-solo) [inline bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch:383]: TODO remove this

File: bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch [hunk 3]
```diff
@@ -0,0 +1,3422 @@
...
+ } // namespace Common
+ } // namespace Filters
+ } // namespace Extensions
+-} // namespace Envoy
++} // namespace Envoy
+\ No newline at end of file
+diff --git a/source/extensions/filters/http/local_ratelimit/BUILD b/source/extensions/filters/http/local_ratelimit/BUILD
+index 88d4042e98..d4ed3daa7e 100644
+--- source/extensions/filters/http/local_ratelimit/BUILD
++++ source/extensions/filters/http/local_ratelimit/BUILD
```
Comment (ashishb-solo) [inline bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch:1123]: nit: it looks like we can remove this file from the patch

File: bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch [hunk 4]
```diff
@@ -0,0 +1,3422 @@
...
+@@ -44,4 +45,4 @@ LEGACY_REGISTER_FACTORY(LocalRateLimitFilterConfig,
+ } // namespace LocalRateLimitFilter
+ } // namespace HttpFilters
+ } // namespace Extensions
+-} // namespace Envoy
++} // namespace Envoy
+\ No newline at end of file
+diff --git a/source/extensions/filters/http/local_ratelimit/config.h b/source/extensions/filters/http/local_ratelimit/config.h
+index 96ae7caa5b..b009c9bcdb 100644
+--- source/extensions/filters/http/local_ratelimit/config.h
```
Comment (ashishb-solo) [inline bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch:1167]: nit: we can probably remove this from the patch as well

**Review by ashishb-solo — COMMENTED** [review]
Looking good to me. I don't see any major issues and found a couple tiny nits.

My biggest question at the moment is what the role of `lbucket.iff` is. Is this something that we need in this patch/branch? Was it added by mistake?

For future reference, it looks like the majority of the complexity in the diffs here lies in these two files:

```
source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc
source/extensions/filters/http/local_ratelimit/local_ratelimit.cc
```

(and their associated headers too, of course).

File: bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch [hunk 5]
```diff
@@ -0,0 +1,3422 @@
...
++  }
++
++  /**
++   * Local descriptor map.
++   */
++  template <class V> using Map = absl::flat_hash_map<LocalDescriptor, V, Hash, Equal>;
+ };
+ 
+ /*
+diff --git a/lbucket.iff b/lbucket.iff
```
Comment (ashishb-solo) [inline bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch:112]: i'm a bit unclear on what this file is doing in the pull request and what its role is. was this perhaps added by mistake?

File: bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch [hunk 6]
```diff
@@ -0,0 +1,3422 @@
...
+@@ -94,7 +94,7 @@ LocalRateLimitStats Config::generateStats(const std::string& prefix, Stats::Scop
+   return {ALL_LOCAL_RATE_LIMIT_STATS(POOL_COUNTER_PREFIX(scope, final_prefix))};
+ }
+ 
+-bool Config::canCreateConnection() { return rate_limiter_->requestAllowed({}); }
++bool Config::canCreateConnection() { return rate_limiter_->requestAllowed({}).allowed; }
+ 
+ Network::FilterStatus Filter::onNewConnection() {
+   if (!config_->enabled()) {
+diff --git a/test/common/common/BUILD b/test/common/common/BUILD
```
Comment (ashishb-solo) [inline bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch:1579]: the substantial amount of changes in this patch, i would really really like to see the tests compile and passing (as mentioned [here](https://github.com/solo-io/envoy-gloo/pull/357#issuecomment-2253175286)), if that's something that you're able to pull off.

that said, it does look like everything is passing [here](https://github.com/solo-io/gloo/pull/9822), so if we can't get the envoy tests merged and passing here, i think we should be good enough

Comment (ashishb-solo) [inline bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch:1579]: putting a comment here to act as a placeholder of whether we should do some sort of performance testing before merging/releasing this (especially considering that we are backporting this large change on our LTS branches)

**Review by ashishb-solo — COMMENTED** [review]
added a few more comments surrounding testing

File: bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch (hunk 1 above)
Comment (nfuden) [reply bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch:375]: I explicitly want this left in as it is what the actual upstream version will have.

File: bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch (hunk 2 above)
Comment (nfuden) [reply bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch:383]: This is a reminder that this patch needs to be removed when we upgrade ourselves to 1.31

File: bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch (hunk 5 above)
Comment (ashishb-solo) [reply bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch]: file was committed by accident

File: ci/cloudbuild.yaml [hunk 7]
```diff
@@ -12,10 +12,9 @@ steps:
   secretEnv:
   - 'GCP_SERVICE_ACCOUNT_KEY'
 
-- name: 'envoyproxy/envoy-build-ubuntu:41c5a05d708972d703661b702a63ef5060125c33'
```
Comment (ashishb-solo) [inline ci/cloudbuild.yaml:15]: why was static analysis removed?

File: ci/cloudbuild.yaml [hunk 8]
```diff
@@ -12,10 +12,9 @@ steps:
...
-  id: 'static_analysis'
-  allowFailure: true
-  args: ['ci/static_analysis.sh']
+- name: 'envoyproxy/envoy-build-ubuntu:f94a38f62220a2b017878b790b6ea98a0f6c5f9c'
+  id: 'do_upstream_ci'
+  args: ['ci/do_ci.sh', 'dev','@envoy//test/extensions/filters/common/local_ratelimit:local_ratelimit_test', '@envoy//test/extensions/filters/http/local_ratelimit/...']
```
Comment (ashishb-solo) [inline ci/cloudbuild.yaml:17]: could you comment here that these tests are here to verify that patch 0004-local-rate-limit....patch works successfully?

and could you also document/comment on the other tests from the other patch that were not included here, ie. whether they pass locally and/or in ci and why?

File: ci/cloudbuild.yaml (hunk 7 above)
Comment (nfuden) [reply ci/cloudbuild.yaml:15]: because its in a github action already

File: ci/cloudbuild.yaml (hunk 8 above)
Comment (nfuden) [reply ci/cloudbuild.yaml:17]: what do you mean by other tests from other patch?

Comment (ashishb-solo) [reply ci/cloudbuild.yaml]: sorry for the typo, i meant the tests listed in this comment: https://github.com/solo-io/envoy-gloo/pull/357#issuecomment-2253175286 (which i can see you already addded)

**Review by ashishb-solo — APPROVED** [review]
looks great! i really like having additional actions like [this](https://github.com/solo-io/envoy-gloo/pull/357/files#diff-36c0eaa91f3225e74aa65aea9cead270644116726f4c6488a516c206fba6d094R15-R19) to test the veracity of the upstream patches that we backport

**Pull request merged**